src/coreclr/tools/aot/ILCompiler.Compiler/Compiler/SubstitutedILProvider.cs

```diff
@@ -906,9 +927,104 @@ private static bool TryExpandTypeEquality(MethodIL methodIL, byte[] body, Opcode
 
             constant = equality.Value ? 1 : 0;
 
-            if (op == "op_Inequality")
-                constant ^= 1;
+            return true;
+        }
+
+        private bool TryExpandTypeEquality_TokenOther(MethodIL methodIL, byte[] body, OpcodeFlags[] flags, int offset, int ldInstructionSize, bool expectGetType, out int constant)
+        {
+            // We expect to see a sequence:
+            // ldtoken Foo
+            // call GetTypeFromHandle
+            // ldloc.X/ldloc_s X/ldarg.X/ldarg_s X
+            // [optional] call Object.GetType
+            // -> offset points here
+            //
+            // The ldtoken part can potentially be in the second argument position
+
+            constant = 0;
+            int sequenceLength = 5 + 5 + ldInstructionSize + (expectGetType ? 5 : 0);
+            if (offset < sequenceLength)
+                return false;
+
+            if ((flags[offset - sequenceLength] & OpcodeFlags.InstructionStart) == 0)
+                return false;
+
+            ILReader reader = new ILReader(body, offset - sequenceLength);
+
+            TypeDesc knownType = null;
+
+            // Is the ldtoken in the first position?
+            if (reader.PeekILOpcode() == ILOpcode.ldtoken)
+            {
+                knownType = ReadLdToken(ref reader, methodIL, flags);
+                if (knownType == null)
+                    return false;
+
+                if (!ReadGetTypeFromHandle(ref reader, methodIL, flags))
+                    return false;
+            }
+
+            ILOpcode opcode = reader.ReadILOpcode();
+            if (ldInstructionSize == 1 && opcode is (>= ILOpcode.ldloc_0 and <= ILOpcode.ldloc_3) or (>= ILOpcode.ldarg_0 and <= ILOpcode.ldarg_3))
+            {
+                // Nothing to read
+            }
+            else if (ldInstructionSize == 2 && opcode is ILOpcode.ldloc_s or ILOpcode.ldarg_s)
+            {
+                reader.ReadILByte();
+            }
+            else if (ldInstructionSize == 3 && opcode is ILOpcode.ldloc or ILOpcode.ldarg)
+            {
+                reader.ReadILUInt16();
+            }
+            else
+            {
+                return false;
+            }
+
+            if ((flags[reader.Offset] & OpcodeFlags.BasicBlockStart) != 0)
+                return false;
+
+            if (expectGetType)
+            {
+                if (reader.ReadILOpcode() is not ILOpcode.callvirt and not ILOpcode.call)
+                    return false;
 
+                // We don't actually mind if this is not Object.GetType
```

Comment:
If it is an arbitrary call, can it return a type that happens to be equal to the other type?

Or is the idea that this case will fail the `CanReferenceConstructedTypeOrCanonicalFormOfType` check below? Ie the other argument can be anything. We are just skipping the specific common patterns here to keep things simple.

Reply:
Yes, we should be okay with any value loaded from a local or parameter. So also any value a method call could return.

We just don't have facilities to accept any value, so only a couple recognized patterns are allowed. Allowing any instance method call is less work than also checking if it's `object.GetType`.